src/main/java/net/openhft/chronicle/network/tcp/FastJ8SocketChannel.java

```diff
@@ -99,17 +100,27 @@ public int write(ByteBuffer byteBuffer) throws IOException {
     }
 
     int readInternal(ByteBuffer buf) throws IOException {
-        int n = OS.read0(fd, IOTools.addressFor(buf) + buf.position(), buf.remaining());
-        if ((n == IOTools.IOSTATUS_INTERRUPTED) && socketChannel.isOpen()) {
-            // The system call was interrupted but the channel
-            // is still open, so retry
-            return 0;
+        try {
+            int n = OS.read0(fd, IOTools.addressFor(buf) + buf.position(), buf.remaining());
+            if ((n == IOTools.IOSTATUS_INTERRUPTED) && socketChannel.isOpen()) {
+                // The system call was interrupted but the channel
+                // is still open, so retry
+                return 0;
+            }
+            int ret = IOTools.normaliseIOStatus(n);
+            if (ret > 0)
+                buf.position(buf.position() + ret);
+            else if (ret < 0)
+                open = false;
+            return ret;
+        } catch (IOException e) {
+            if (!super.isOpen()) {
+                open = false;
+                final ClosedChannelException closedChannelException = new ClosedChannelException();
+                closedChannelException.initCause(e);
+                throw closedChannelException;
+            }
+            throw e;
```

Comment:
could make this try/catch block more localised? I don't know if anything else in there throws an IOException